### PR TITLE
feat: Page.routeFromHar

### PR DESCRIPTION
**Pull request by yury-s**

### CODE REVIEW

File: docs/src/api/class-browsercontext.md [hunk 1]
```diff
@@ -1025,6 +1025,37 @@ handler function to route the request.
 
 How often a route should be used. By default it will be used every time.
 
+## async method: BrowserContext.routeFromHar
+
+Provides the capability to serve network requests that are made in the context from prerecorded HAR file.
+
+If HAR file contains an entry with the matching url and HTTP method, then the entry's headers, status and body will be used to fulfill. An entry resulting in a redirect will be followed automatically. If there is no matching entry in the file the execution continues to try other configured HAR files and [Route] handlers.
+If `path` is a relative path, then it is resolved relative to the current working directory.
```
Comment (dgozman) [inline docs/src/api/class-browsercontext.md:1033]: Move this sentence to `harPath` instead.

File: docs/src/api/class-browsercontext.md [hunk 2]
```diff
@@ -1025,6 +1025,37 @@ handler function to route the request.
...
+
+Path to the HAR file with prerecorded network data.
+
+### option: BrowserContext.routeFromHar.strict
+- `strict` <[boolean]>
+
+If set to true any request not found in the HAR file will be aborted. If set to
+false missing requests will continue normal flow and can be handled by other
+[Route] handlers or served from other HAR files configured with [`method: BrowserContext.routeFromHar`].
+Defaults to false.
```
Comment (dgozman) [inline docs/src/api/class-browsercontext.md:1050]: I thought we agreed to default to `true`?

File: docs/src/api/class-page.md [hunk 3]
```diff
@@ -2732,6 +2732,37 @@ handler function to route the request.
 
 How often a route should be used. By default it will be used every time.
 
+## async method: Page.routeFromHar
+
+Provides the capability to serve network requests that are made by a page from prerecorded HAR file.
+
+If HAR file contains an entry with the matching url and HTTP method, then the entry's headers, status and body will be used to fulfill. An entry resulting in a redirect will be followed automatically. If there is no matching entry in the file the execution continues to try other configured HAR files and [Route] handlers.
+If `path` is a relative path, then it is resolved relative to the current working directory.
```
Comment (dgozman) [inline docs/src/api/class-page.md:2740]: Same here.

File: packages/playwright-core/src/client/harRouter.ts [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+import type { Route } from './network';
+import type { Page } from './page';
+
+type HarHandler = {
+  pattern: string | RegExp;
+  handler: (route: Route) => any;
+};
+
+export class HarRouter {
+  private _harPathToHander: Map<string, HarHandler> = new Map();
```
Comment (dgozman) [inline packages/playwright-core/src/client/harRouter.ts:29]: typo: `_harPathToHandler`

File: packages/playwright-core/src/client/harRouter.ts [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+  private readonly owner: BrowserContext | Page;
+
+  constructor(owner: BrowserContext | Page) {
+    this.owner = owner;
+  }
+
+  async routeFromHar(path: string, options?: { strict?: boolean; url?: string|RegExp; }): Promise<void> {
+    const harFile = JSON.parse(await fs.promises.readFile(path, 'utf-8')) as HARFile;
+    if (this._harPathToHander.has(path))
+      await this.unrouteFromHar(path);
```
Comment (dgozman) [inline packages/playwright-core/src/client/harRouter.ts:39]: Why this one? Can I have a single HAR serving two url patterns?

**Review by dgozman — APPROVED** [review]

File: docs/src/api/class-browsercontext.md (hunk 2 above)
Comment (yury-s) [reply docs/src/api/class-browsercontext.md:1050]: Let's double check tomorrow in the meeting.

**Pull request merged**